torch/distributed/_spmd/graph_optimization.py

```diff
@@ -0,0 +1,408 @@
+# Owner(s): ["oncall: distributed"]
+import collections
+import logging
+import operator
+import tempfile
+from dataclasses import dataclass
+from functools import wraps
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+import torch
+import torch.distributed as dist
+import torch.fx as fx
+from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
+from torch.distributed._spmd.graph_utils import (
+    CommType,
+    dump_graphs_to_files,
+    get_output,
+)
+from torch.distributed._spmd.iter_graph_module import IterGraphModule
+from torch.fx.passes.shape_prop import TensorMetadata
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+logger: logging.Logger = logging.getLogger("graph_optimization")
+aten = torch.ops.aten
+fake_tensor_mode = FakeTensorMode()
+
+_optimized_func: Set[str] = set()
+_run_before_sets: DefaultDict[str, Set[str]] = collections.defaultdict(set)
+_dump_graph_folder: str = ""
+
+
+logger: logging.Logger = logging.getLogger("graph_optimization")
+
+
+def enable_graph_optimization_dump(folder: str = ""):
+    global _dump_graph_folder
+    if not folder:
+        folder = tempfile.mkdtemp()
+    _dump_graph_folder = folder
+
+
+def graph_optimization_pass(
+    run_after: Iterable[str] = tuple(),
+) -> Callable[..., Callable[..., None]]:
+    """
+    The contract of graph optimization pass. All the passes should be wrapped with
+    this decorator.
+    """
+
+    def inner(
+        func: Callable[..., Union[fx.GraphModule, IterGraphModule]]
+    ) -> Callable[..., Union[fx.GraphModule, IterGraphModule]]:
+        for name in run_after:
+            _run_before_sets[name].add(func.__name__)
+
+        @wraps(func)
+        def pass_wrapper(
+            gm: Union[fx.GraphModule, IterGraphModule], *args: Any, **kwargs: Any
+        ) -> None:
+            assert isinstance(
+                gm, (fx.GraphModule, IterGraphModule)
+            ), "The first argument of the pass must be either fx.GraphModule or IterGraphModule"
+            assert (
+                func.__name__ not in _optimized_func
```

Comment:
this seems to be brittle, as functions with the same name would collide even if they are from different packages?

Reply:
The very original design is to put all the optimization passes in a class. This way, the name will guarantee to be unique. But I feel it is a little bit overkill to use another class. Any suggestion of this?

Reply:
On second thought, the design should be fine.A pass will only be added to `_optimized_func` if the pass is wrapped with `graph_optimization_pass` and run. If there are more than one optimization passes that have the same naming, I feel this is actually good to error out because the same naming will eventually cause problems.

Reply:
> Any suggestion of this?

Maybe this?

```python
".".join([func.__module__, func.__name__])
```